accepted/exception-handling.md

```diff
@@ -0,0 +1,553 @@
+      bespoke unwinder for just the subset of DWARF that Cranelift
+      emits, or
+
+   b. the side table can be a completely bespoke format.
+
+   This approach is zero-cost for the non-exceptional path.
+
+   This approach has some complexity-at-a-distance, as the compiler
+   needs to correctly emit unwind tables and the runtime needs to
+   correctly interpret the tables during unwinding.
+
+2. **Calling convention**: we can extend the calling convention to
+   propagate exception information between function calls, for example
+   in a dedicated register (riscv64) or a particular flags bit
+   (aarch64, x86\_64, and I think s390x). If that register is non-zero
+   or the bit is set, then this is an exceptional return, rather than
+   a normal return, and we would branch directly to the `exn_label`
+   landing pad.
+
+   This approach imposes overhead on the non-exceptional path, even
+   when the program never throws an exception.
+
+   Correctness does not involve any coordination between distant
+   components (such as the runtime's interpretation of unwind tables
+   and the compiler) since the compiler can emit everything necessary
+   for correctness inline.
+
+Because the calling-convention approach imposes overhead on the
+non-exceptional path, we could never turn Wasm exceptions on by
+default in Wasmtime using it. That means that we would need to support
+two different calling conventions (with- and without exceptions) for
+every target indefinitely. And these calling conventions wouldn't just
+have a different set of caller- vs. callee-saved registers or argument
+and return value passing locations; they would have all this extra
+check-for-and-branch-on-exception code to maintain. We consider this
+scenario a non-starter. Furthermore, it doesn't make sense to use the
+calling-convention approach as an incremental milestone on the way to
+a side-table implementation, since the side-table implementation would
+not build on top of the calling-convention implementation and would
+only throw away work we previously did. We could have, instead,
+dedicated that effort towards implementing the final side-table
+approach in the first place. Therefore, **we will implement the
+side-table approach** and we will do so right from the beginning.
+
+We are now left with the question of whether to use side table option
+(a) a DWARF subset or (b) a bespoke format. Let's start with some
+observations:
+
+* For the foreseeable moment, we do not plan to remove or replace
+  Wasmtime's frame-pointer-based stack walking as part of implementing
+  the Wasm exceptions proposal; unwinding exceptions will use frame
+  pointers to walk the stack, the same way that existing stack walking
+  happens today.[^fp-unwind-future]
+
+* Again, at least for the foreseeable moment, we can make
+  `try_call[_indirect]` instructions use a variant of the target's
+  Wasm calling convention where all registers are caller-saved.
```

Comment:
This is only done for the tail calling convention. For the system_v calling convention try_call uses the exact same set of clobbered registers as regular call.